src/chainstate/stacks/index/node.rs

```diff
@@ -0,0 +1,2598 @@
+    pub const Empty : u8 = 0;
+    pub const Leaf: u8 = 1;
+    pub const Node4 : u8 = 2;
+    pub const Node16 : u8 = 3;
+    pub const Node48 : u8 = 4;
+    pub const Node256 : u8 = 5;
+}
+
+/// A node ID encodes a back-pointer if its high bit is set
+pub fn is_backptr(id: u8) -> bool {
+    id & 0x80 != 0
+}
+
+/// Set the back-pointer bit
+pub fn set_backptr(id: u8) -> u8 {
+    id | 0x80
+}
+
+/// Clear the back-pointer bit
+pub fn clear_backptr(id: u8) -> u8 {
+    id & 0x7f
+}
+
+/// A path in the Trie is the SHA2-512/256 hash of its key.
+pub struct TriePath([u8; 32]);
+impl_array_newtype!(TriePath, u8, 32);
+impl_array_hexstring_fmt!(TriePath);
+impl_byte_array_newtype!(TriePath, u8, 32);
+
+pub const TRIEPATH_MAX_LEN : usize = 32;
+
+impl TriePath {
+    pub fn from_key(k: &String) -> TriePath {
+        let h = TrieHash::from_data(k.as_bytes());
+        let mut hb = [0u8; TRIEPATH_MAX_LEN];
+        hb.copy_from_slice(h.as_bytes());
+        TriePath(hb)
+    }
+}
+
+/// All Trie nodes implement the following methods:
+pub trait TrieNode {
+    /// Node ID for encoding/decoding
+    fn id(&self) -> u8;
+
+    /// Is the node devoid of children?
+    fn empty() -> Self;
+
+    /// Follow a path character to a child pointer
+    fn walk(&self, chr: u8) -> Option<TriePtr>;
+
+    /// Insert a child pointer if the path character slot is not occupied.
+    /// Return true if inserted, false if the slot is already filled
+    fn insert(&mut self, ptr: &TriePtr) -> bool;
+
+    /// Replace an existing child pointer with a new one.  Returns true if replaced; false if the
+    /// child does not exist.
+    fn replace(&mut self, ptr: &TriePtr) -> bool;
+
+    /// Read an encoded instance of this node from a byte stream and instantiate it.
+    fn from_bytes<R: Read>(r: &mut R) -> Result<Self, Error>
+        where Self: std::marker::Sized;
+
+    /// Encode this node instance into a byte stream and write it to the given memory buffer.
+    fn to_bytes(&self, buf: &mut Vec<u8>) -> ();
+
+    /// Encode the consensus-relevant bytes of this node and write it to the given memory buffer.
+    fn to_consensus_bytes(&self, &mut Vec<u8>) -> ();
+
+    /// Calculate how many bytes this node will take to encode.
+    fn byte_len(&self) -> usize;
+
+    /// Get a reference to the children of this node.
+    fn ptrs(&self) -> &[TriePtr];
+
+    // this is a way to construct a TrieNodeType from an object that implements this trait
+    // DO NOT USE DIRECTLY
+    fn try_as_node4(&self) -> Option<TrieNodeType>;
+    fn try_as_node16(&self) -> Option<TrieNodeType>;
+    fn try_as_node48(&self) -> Option<TrieNodeType>;
+    fn try_as_node256(&self) -> Option<TrieNodeType>;
+    fn try_as_leaf(&self) -> Option<TrieNodeType>;
+}
+
+/// Child pointer
+#[derive(Debug, Clone, Copy, PartialEq)]
+pub struct TriePtr {
+    pub id: u8,             // ID of the child.  Will have bit 0x80 set if the child is a back-pointer (in which case, back_block will be nonzero)
```

Comment:
this is more of a type than an ID, correct?

the u8 corresponds to the pair `(TrieNodeType, isBackPointer)`, right?

Reply:
Yes, that is correct.  Not sure, semantically, what the difference is between an ID and a type.  Do you want me to change it?

Reply:
No real need to change, but was just a little confused because usually an identifier refers to a unique objects while types refer to a class of things.